crates/llms/src/anthropic/chat.rs

```diff
@@ -0,0 +1,397 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#![allow(deprecated)]
+#![allow(clippy::missing_errors_doc)]
+use std::pin::Pin;
+use std::time::SystemTime;
+
+use crate::chat::nsql::SqlGeneration;
+use crate::chat::Chat;
+use async_openai::error::{ApiError, OpenAIError};
+use async_openai::types::{
+    ChatChoice, ChatCompletionMessageToolCall, ChatCompletionNamedToolChoice,
+    ChatCompletionRequestAssistantMessage, ChatCompletionRequestAssistantMessageContent,
+    ChatCompletionRequestAssistantMessageContentPart, ChatCompletionRequestMessage,
+    ChatCompletionRequestMessageContentPartText, ChatCompletionRequestSystemMessage,
+    ChatCompletionRequestSystemMessageContent, ChatCompletionRequestSystemMessageContentPart,
+    ChatCompletionRequestToolMessage, ChatCompletionRequestToolMessageContent,
+    ChatCompletionRequestToolMessageContentPart, ChatCompletionRequestUserMessage,
+    ChatCompletionRequestUserMessageContent, ChatCompletionRequestUserMessageContentPart,
+    ChatCompletionResponseMessage, ChatCompletionResponseStream, ChatCompletionToolChoiceOption,
+    ChatCompletionToolType, CompletionUsage, CreateChatCompletionRequest,
+    CreateChatCompletionResponse, FinishReason, FunctionCall, FunctionName, Role, Stop,
+};
+
+use super::types::{
+    AnthropicModelVariant, ContentBlock, ContentParam, MessageCreateParams, MessageCreateResponse,
+    MessageParam, MessageRole, MetadataParam, ResponseContentBlock, StopReason, TextBlockParam,
+    ToolChoiceParam, ToolResultBlockParam, ToolUseBlockParam,
+};
+use super::types_stream::{transform_stream, AnthropicStreamError, MessageCreateStreamResponse};
+use super::Anthropic;
+use async_trait::async_trait;
+use futures::Stream;
+
+#[async_trait]
+impl Chat for Anthropic {
+    fn as_sql(&self) -> Option<&dyn SqlGeneration> {
+        None
+    }
+
+    async fn chat_stream(
+        &self,
+        req: CreateChatCompletionRequest,
+    ) -> Result<ChatCompletionResponseStream, OpenAIError> {
+        if req.stream.is_some_and(|b| !b) {
+            return Err(OpenAIError::InvalidArgument(
+                "When stream is false, use Chat::create".into(),
+            ));
+        }
```

Comment:
This smells like an architectural issue with the code - it should be disallowed by the type system to call this method if stream is false.

Reply:
This is not needed, copied directly from `async-openai` that doesn't make it impossible. Removed.